Migrate `data` field of `ScheduledEmail` class from `TextField` to `JSONField`

edith007 opened the pull request:
The PR aims to migrate the data field of ScheduledEmail model from TextField to JSONField.
This was done as part of the audit which was mentioned in this issue: https://github.com/zulip/zulip/issues/17205

The changes done are:
- Migrate the data field from TextField to JSONField.
- Backfill migration to copy the elements in the field to their new type.
- Make changes across the codebase to use `data` as an JSONField.

**Testing Plan**:
- I have also tested it in the shell. I printed the content of the data field and the contents of the data_json field, and verified manually that the contents are same and the content of the data_json are deserialzed JSON stored in a dictionary.

edith007 committed:
data: Apply changes to use data field as an JSONField.

The data field is migrated to JSONField. This commit makes
changes across the codebase to use the field as an JSONField and
ensures that all the tests pass and the behaviour is preseved.

timabbott commented on zerver/migrations/0324_backfill_scheduledemail_data_to_jsonfield.py:
```diff
@@ -0,0 +1,27 @@
+
+    for scheduledemail in ScheduledEmail.objects.filter(data__isnull=False):
+        data_json = orjson.loads(scheduledemail.data)
+        scheduledemail.data_json = dict(data_json)
```
Why this `dict()` line here?

edith007 replied:
I wanted to copy the `data_json` dictionary to `data_json` field of the model, that is why I used `dict(data_json)` .. but now that you have mentioned, I realize that I should have done
`scheduleemail.data_json = data_json` ... will update it!

timabbott commented:
This looks reasonable to me in terms of the final state (there may be follow-up cleanups we can do, but that's not critical for doing the migration).  The main question I have is how we feel about this migration sequence in terms of its ability to be deployed in production.  @alexmv @mateuszmandera thoughts? 



See also https://github.com/zulip/zulip/pull/18146, a parallel PR.

edith007 committed:
migration: Migrate data field of ScheduledEmail model to JSONField.

Migrate the data field of ScheduledEmail model to JSONField. The
migration is done through a raw SQL query which changes the type of the
field from TextField to JSONField and also converts the type of the
data stored in that field.

edith007 commented:
@timabbott @mateuszmandera 
updated the migration to use raw SQL to migrate the data field from TextField to JSONField.

I tested the migration by doing the following in shell:
-- This is the content of the data field before the migration is applied:
![image1](https://user-images.githubusercontent.com/53316982/117058167-ed3ff700-ad3b-11eb-92b8-87d86f043efd.png)

-- This is the content of the data field after the migration is applied:
![image2](https://user-images.githubusercontent.com/53316982/117058224-0052c700-ad3c-11eb-9ec4-e70b123d10de.png)

I have also ensured all the backend tests pass.
